### PR TITLE
A2-10-1: add functions and types to identifier consideration

### LINKED ISSUE #118: `A2-10-1`: Report type and function identifier hiding
### Affected rules
 - `A2-10-1`

### Description

`A2-10-1` states "An identifier declared in an inner scope shall not hide an identifier declared in an outer scope.". The existing query looks for user defined variables which cause hiding, but functions and types can also be hidden (and cause hiding). 

### Example
```cpp
int a;
namespace b {
        void a() { } // NON_COMPLIANT
        int c() { return a; }
}
```

### PR DESCRIPTION
**Pull request by knewbury01**
## Description

fixes #118 

the query seems to maybe be slower than before, will need this confirmed

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - A2-10-1
     - RULE-5-3

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Review by rvermeulen — CHANGES_REQUESTED** [review]
The generalization of identifiers needs some further consideration, because of the increased complexity and incomplete exclusions part of the first iteration.

Possible FN are class members where a derived class member hides a base class member. It is possibly because the outer/inner scope definition for this rule doesn't mention it, but it successor in Misra C++ 23 does include that explicitly in the definition of the outer/inner scope.

Our `hides` logic now has to consider functions and their overloads. The query itself uses the strict version that requires a different scope which excludes this as problem. Not sure if we want to correct the non-strict version. 

Given that the Misra C/C++ 23 version of this rule, rule 6.4.1 and 6.4.2, focuses solely on variables and functions, I wonder if we want to do the same here and exclude types.

File: cpp/common/src/codingstandards/cpp/Scope.qll [hunk 1]
```diff
@@ -204,18 +212,9 @@ class TranslationUnit extends SourceFile {
...
-}
-
-/** Holds if `v2` may hide `v1`. */
-private predicate hides_candidateStrict(UserVariable v1, UserVariable v2) {
-  not v1 = v2 and
-  v2 = getPotentialScopeOfVariableStrict(v1) and
+  v2 = getPotentialScopeOfDeclarationStrict(v1) and
   v1.getName() = v2.getName() and
   // Member variables cannot hide other variables nor be hidden because the can be referenced through their qualified name.
   not (v1.isMember() or v2.isMember()) and
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/Scope.qll:220]: fixed the first one,

for the second one, having a bit of trouble - this wasnt actually an FP , it was an FN
I dont understand lambdas' scopes...
for some reason the body block of a function is not showing up as the ancestor scope of a lambda:
```
void f4() {
  int a1, b;
  auto lambda1 = [a1]() {
    int b = 10; // COMPLIANT - exception - non captured variable b
  };

  auto lambda2 = [b]() {
    int b = 10; // NON_COMPLIANT[FALSE_NEGATIVE] - not an exception - captured variable b
  };
}
```
`s.getAnAncestor+()` ([from here](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/Scope.qll#L148)) quick eval'd

shows  - 
```

{ ... } | decltype([...](...){...})
-- | --
43 | { ... } | operator()
44 | { ... } | (unnamed constructor)
45 | { ... } | decltype([...](...){...})


```
none of which are the f4 function body or anything else upwards of that

**Comment by knewbury01** [comment]
> The generalization of identifiers needs some further consideration, because of the increased complexity and incomplete exclusions part of the first iteration.
> 
> Possible FN are class members where a derived class member hides a base class member. It is possibly because the outer/inner scope definition for this rule doesn't mention it, but it successor in Misra C++ 23 does include that explicitly in the definition of the outer/inner scope.

this is covered in  [A10-2-1](https://github.com/github/codeql-coding-standards/blob/main/cpp/autosar/src/rules/A10-2-1/NonVirtualPublicOrProtectedFunctionsRedefined.ql) - so we should be ok to not worry about this exact case? 

> 
> Our `hides` logic now has to consider functions and their overloads. The query itself uses the strict version that requires a different scope which excludes this as problem. Not sure if we want to correct the non-strict version.
> 

currently the only use of `hides` is [here](https://github.com/github/codeql-coding-standards/blob/main/cpp/cert/src/rules/DCL53-CPP/LocalConstructorInitializedObjectHidesIdentifier.ql#L23), where that wasnt relevant, but I omitted the overloads nonetheless, assuming that overloads are always intentional (not hiding), which I think sounds reasonable?

> Given that the Misra C/C++ 23 version of this rule, rule 6.4.1 and 6.4.2, focuses solely on variables and functions, I wonder if we want to do the same here and exclude types.

done, great catch (didnt own the 23 version til this comment)

File: cpp/common/src/codingstandards/cpp/Scope.qll (hunk 1 above)
[the remark this answers is not in the document]
Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/Scope.qll:222]: nevermind, was able to write something to cover both cases!

File: cpp/common/src/codingstandards/cpp/Scope.qll [hunk 2]
```diff
@@ -57,10 +57,18 @@ private Element getParentScope(Element e) {
...
+
+/** A construct which is defined by the user, rather than being from a third party or compiler generated. */
+class UserDeclaration extends Declaration {
+  UserDeclaration() {
     exists(getFile().getRelativePath()) and
-    not isCompilerGenerated() and
+    not this.(Variable).isCompilerGenerated() and
+    not this.(Function).isCompilerGenerated() and
     not this.(Parameter).getFunction().isCompilerGenerated() and
+    // will falsely conflict
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/Scope.qll:70]:
```suggestion
    // Class template instantiations are compiler generated instances that share the same parent scope. This will result in a cross-product on class template instantiations because they have the same name and same parent scope. We therefore exclude these from consideration like we do with other compiler generated identifiers of interest.
```

File: cpp/common/src/codingstandards/cpp/Scope.qll [hunk 3]
```diff
@@ -257,20 +267,22 @@ private Stmt getEnclosingStmt(LocalScopeVariable v) {
...
+predicate hides(UserDeclaration v1, UserDeclaration v2) {
   hides_candidate(v1, v2) and
   // Confirm that there's no closer candidate variable which `v2` hides
-  not exists(UserVariable mid |
+  not exists(UserDeclaration mid |
     hides_candidate(v1, mid) and
     hides_candidate(mid, v2)
-  )
+  ) and
+  //ignore intentional overloads
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/Scope.qll:277]:
```suggestion
  // Unlike `hidesStrict`, that requires a different scope, `hides` considers declarations in the same scope. This will include function overloads based on their name. To remove overloads from consideration, we exclude them.
```

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll [hunk 4]
```diff
@@ -11,13 +11,36 @@ abstract class IdentifierHiddenSharedQuery extends Query { }
 
 Query getQuery() { result instanceof IdentifierHiddenSharedQuery }
 
-query predicate problems(UserVariable v2, string message, UserVariable v1, string varName) {
+/**
+ * There is a lambda that contains a declaration
+ * that hides something that is captured
+ * and the lambda exists in the function where this lamda is enclosed
+ */
+predicate hiddenInLambda(UserDeclaration v2, UserDeclaration v1) {
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll:19]: To make it consistent with `hidesStrict` we should swap the parameters.
`v2` hides `v1`.

```suggestion
predicate hiddenInLambda(UserDeclaration v1, UserDeclaration v2) {
```

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll [hunk 5]
```diff
@@ -11,13 +11,36 @@ abstract class IdentifierHiddenSharedQuery extends Query { }
...
+  exists(Scope s, Closure le |
+    s.getADeclaration() = v2 and
+    s.getAnAncestor() = le and
+    le.getEnclosingFunction().getBasicBlock().(Scope) = v1.getParentScope() and
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll:23]: Why are we looking at the basic block of the function? I believe this works because it coincides with the `BlockStmt` of the function.
Would it be better to look at the transitive parent scope of the closure or does that not have a parent scope that works?

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll (hunk 4 above)
Comment (rvermeulen) [reply cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll:19]: I think overall using of more verbose variables help with understanding the predicate.

```ql
predicate hiddenInLambda(UserDeclaration outerDecl, UserDeclaration innerDecl) {
  exists(Scope scopeOfInnerDecl, Closure closureWithInnerDecl|
    scopeOfInnerDecl.getADeclaration() = innerDecl and
    scopeOfInnerDecl.getAnAncestor() = closureWithInnerDecl and
    closureWithInnerDecl.getEnclosingFunction().getBasicBlock().(Scope) = outerDecl.getParentScope() and
    exists(LambdaCapture cap, Variable v |
      v.getAnAccess() = cap.getInitializer().(VariableAccess) and
      v = outerDecl and
      closureWithInnerDecl.getLambdaExpression().getACapture() = cap
    ) and
    innerDecl.getName() = outerDecl.getName()
  )
}
```

Comment (rvermeulen) [reply cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll:19]: Since we are limiting to variable in the `exists` it would be a performance benefit to change the parameters to `UserVariable` to potentially reduce the number of starting tuples.

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll [hunk 6]
```diff
@@ -11,13 +11,36 @@ abstract class IdentifierHiddenSharedQuery extends Query { }
 
 Query getQuery() { result instanceof IdentifierHiddenSharedQuery }
 
-query predicate problems(UserVariable v2, string message, UserVariable v1, string varName) {
+/**
+ * There is a lambda that contains a declaration
+ * that hides something that is captured
+ * and the lambda exists in the function where this lamda is enclosed
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll:17]:
```suggestion
 * Holds if declaration `v2`, declared in a lambda, hides a declaration `v1` captured by the lambda.
```

Not that this is based on the current version. If the suggestions below are considered, we need to adjust this accordingly.

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll [hunk 7]
```diff
@@ -11,13 +11,36 @@ abstract class IdentifierHiddenSharedQuery extends Query { }
...
+}
+
+query predicate problems(UserDeclaration v2, string message, UserDeclaration v1, string varName) {
   not isExcluded(v1, getQuery()) and
   not isExcluded(v2, getQuery()) and
   //ignore template variables for this rule
   not v1 instanceof TemplateVariable and
   not v2 instanceof TemplateVariable and
-  hidesStrict(v1, v2) and
+  //ignore types for this rule
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll:39]:
```suggestion
  //ignore types for this rule
```
Can we also document why for future reference?

File: cpp/common/src/codingstandards/cpp/Scope.qll [hunk 8]
```diff
@@ -287,3 +299,18 @@ predicate hasClassScope(Declaration decl) { exists(decl.getDeclaringType()) }
 
 /** Holds if `decl` has block scope. */
 predicate hasBlockScope(Declaration decl) { exists(BlockStmt b | b.getADeclaration() = decl) }
+
+/**
+ * identifiers in nested (named/nonglobal) namespaces are exceptions to hiding due to being able access via fully qualified ids
+ */
+predicate excludedViaNestedNamespaces(UserDeclaration v2, UserDeclaration v1) {
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/Scope.qll:306]:
```suggestion
predicate excludedViaNestedNamespaces(UserDeclaration v1, UserDeclaration v2) {
```
Can we use the same order as `hides`, where `v2` hides `v1`.

Better names would be `outerDecl` and `innerDecl` where `innerDecl` hides the `outerDecl`.

**Review by rvermeulen — CHANGES_REQUESTED** [review]
I left a few considerations for the lambda case and proposal for more reorder the parameters for consistency.

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll (hunk 4 above)
Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll:19]: the scope check here is to say - the innerDecl is in a scope inside the lambda, its not for the capture (but this is hopefully more clear now with the var name improvements 😅 )

not sure I understand the case - "The situation becomes more tricky for variables that accessible without a capture." 
if the variable is not captured, then there is no hiding, reference to the compliant exception example in the standard

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll (hunk 5 above)
Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll:23]: "Would it be better to look at the transitive parent scope..."
yeah that sounds more ideal but I wasnt able to get that to work 😞 

quick eval for `le.getParentScope()` has no result

tldr the scopes in lambdas werent as I naively expected, but if you see a better way to do this I could totally be missing something obvious!

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll (hunk 4 above)
Comment (rvermeulen) [reply cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll]: A lambda, can for example refer to a global variable without capturing that.

```
int foo = 0;

int bar() {
    auto lambda_with_shadowing = []() {
        int foo = 1;
        return foo + foo;
    };

    auto lambda_without_shadowing = []() {
        return foo + foo;
    };

    return lambda_with_shadowing();
}
```

I would argue that `lambda_with_shadowing` is non compliant.

Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll]: aha, you are correct ,

I have added cases for the description [here](https://en.cppreference.com/w/cpp/language/lambda) when a lambda can access a variable without explicitly capturing it

File: cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll (hunk 5 above)
Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/rules/identifierhidden/IdentifierHidden.qll]: [your addition](https://github.com/knewbury01/codeql-coding-standards/pull/2) to this model deprecated this comment!

File: cpp/common/src/codingstandards/cpp/Scope.qll [hunk 9]
```diff
@@ -122,31 +132,31 @@ class GeneratedBlockStmt extends BlockStmt {
...
       not result = v and
       v.getName() <= result.getName()
     )
   )
 }
 
-/** Gets a variable that is in the potential scope of variable `v`. */
-private UserVariable getOuterScopesOfVariable_candidate(UserVariable v) {
+/** Gets a Declarationthat is in the potential scope of Declaration `v`. */
+private UserDeclaration getOuterScopesOfDeclaration_candidate(UserDeclaration v) {
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/Scope.qll:153]:
```suggestion
private UserDeclaration getPotentialScopeOfDeclarationStrict_candidate(UserDeclaration v) {
```

I missed this being renamed, but we should keep the previous name.

File: cpp/common/src/codingstandards/cpp/Scope.qll [hunk 10]
```diff
@@ -161,20 +171,20 @@ predicate inSameTranslationUnit(File f1, File f2) {
...
 
 /**
- * Gets a user variable which occurs in the "potential scope" of variable `v`.
+ * Gets a user Declaration which occurs in the "outer scope" of Declaration `v`.
  */
 cached
-UserVariable getPotentialScopeOfVariable(UserVariable v) {
-  result = getPotentialScopeOfVariable_candidate(v) and
+UserDeclaration getPotentialScopeOfDeclarationStrict(UserDeclaration v) {
+  result = getOuterScopesOfDeclaration_candidate(v) and
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/Scope.qll:178]:
```suggestion
  result = getPotentialScopeOfDeclarationStrict_candidate(v) and
```

File: cpp/common/src/codingstandards/cpp/Scope.qll (hunk 9 above)
Comment (rvermeulen) [reply cpp/common/src/codingstandards/cpp/Scope.qll:153]: I noticed the previous name was also outer scope ..., not sure why.

**Comment by rvermeulen** [comment]
With addition of other declarations the performance of this query has significantly decreased. My current hypothesis is that the number of declarations in the global namespace explodes the size of the `getOuterScopesOfDeclaration_candidate` relation that requires some further scrutinizing.

File: cpp/common/src/codingstandards/cpp/Scope.qll (hunk 9 above)
Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/Scope.qll:153]: hmm yep, good catch, also not sure why!

**Review by rvermeulen — APPROVED** [review]
LGTM.
